File: python/python_b2b_seeding/python_service/database/seeds/user_seeder.py
"""
User seeder that creates users, optionally loading from system.admin.seed.json.
Similar to the Node.js UserSeeder implementation.
"""
import json
import os
from typing import Dict
from datetime import datetime

from .base_seeder import BaseSeeder
from database.models.user import User
from database.models.tenant import Tenant


class UserSeeder(BaseSeeder):
    """Seeder for creating users."""
    
    def __init__(self, session, tenant_repository=None, batch_size: int = 1000, verbose: bool = True):
        """
        Initialize user seeder.
        
        Args:
            session: SQLAlchemy session
            tenant_repository: Optional tenant repository (not used in SQLAlchemy version)
            batch_size: Batch size for inserts
            verbose: Whether to print progress
        """
        super().__init__(session, batch_size, verbose)
        self.session = session
    
    def _hash_password(self, password: str) -> str:
        """Hash password using bcrypt."""
        try:
            import bcrypt
            return bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
        except ImportError:
            # Fallback for development
            import hashlib
            return hashlib.sha256(password.encode()).hexdigest()
    
    def _load_seed_data(self) -> Dict:
        """Load seed data from system.admin.seed.json."""
        primary_seed_path = os.path.join(os.getcwd(), 'seed.data', 'system.admin.seed.json')
        
        try:
            if os.path.exists(primary_seed_path):
                with open(primary_seed_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    return {
                        'username': data.get('UserName', 'admin'),
                        'firstName': data.get('FirstName', 'System'),
                        'lastName': 'Admin',
                        'email': data.get('Email', 'sys.admin@example.com'),
                        'password': data.get('Password', 'ChangeMe123!'),
                        'phoneNumber': data.get('PhoneNumber', None),
                        'phoneCode': data.get('PhoneCode', '+91'),
                    }
        except Exception as e:
            if self.verbose:
                print(f"[WARN] Could not load seed data: {e}")
        
        return None
    
    def seed(self, count: int = 2, session=None) -> int:
        """
        Seed users.
        
        Args:
            count: Number of users to create (default: 2)
            session: SQLAlchemy session
            
        Returns:
            Number of users created
        """
        if session is None:
            session = self.session
            
        print(f"Seeding {count} users...")
        
        # Get all tenants
        tenants = session.query(Tenant).all()
        
        if len(tenants) > 0:
            print(f"[INFO] Assigning users to {len(tenants)} tenant(s)")
        else:
            print("[WARN] No tenants found. Users will be created without tenantId.")
        
        # Load seed data
        seed_data = self._load_seed_data()
        
        default_password = self._hash_password('password123')
        
        # Default values
        usernames = ['john.smith', 'sarah.brown', 'mike.johnson', 'jane.williams', 'david.davis']
        first_names = ['John', 'Jane', 'Mike', 'Sarah', 'David']
        last_names = ['Smith', 'Johnson', 'Williams', 'Brown', 'Davis']
        email_domain = 'example.com'
        
        # Override with seed data if available
        if seed_data:
            if seed_data.get('username'):
                usernames = [seed_data['username']]
            if seed_data.get('firstName'):
                first_names = [seed_data['firstName']]
            if seed_data.get('email') and '@' in seed_data['email']:
                email_domain = seed_data['email'].split('@')[1]
        
        users = []
        
        for i in range(count):
            # Use seed data for first user if available
            if i == 0 and seed_data:
                user = {
                    'username': seed_data.get('username', f'user{i}'),
                    'email': seed_data.get('email', f'user{i}@{email_domain}'),
                    'password': self._hash_password(seed_data.get('password', 'ChangeMe123!')),
                    'first_name': seed_data.get('firstName', self.get_random_choice(first_names)),
                    'last_name': seed_data.get('lastName', 'Admin'),
                    'is_active': True,
                    'is_email_verified': True,
                    'is_phone_verified': False,
                }
                
                # Add phone number if available
                if seed_data.get('phoneNumber'):
                    user['phone_number'] = seed_data['phoneNumber']
                if seed_data.get('phoneCode'):
                    user['country_code'] = seed_data['phoneCode']
            else:
                user = {
                    'username': self.get_random_choice(usernames) if i < len(usernames) else f'user{i}',
                    'email': f'user{i}@{email_domain}',
                    'password': default_password,
                    'first_name': self.get_random_choice(first_names),
                    'last_name': self.get_random_choice(last_names),
                    'is_active': self.get_random_boolean(0.9),
                    'is_email_verified': self.get_random_boolean(0.8),
                    'is_phone_verified': self.get_random_boolean(0.5),
                }
            
            # Assign tenant if available
            if tenants and len(tenants) > 0:
                tenant = self.get_random_choice(tenants)
                user['tenant_id'] = tenant.id
            
            users.append(user)
        
        from database.models.user import User
        self.batch_insert(users, session, model_class=User)
        print(f"Successfully seeded {self.created_count} users")
        return self.created_count

File: dotnet/dotnet_b2b_seeding/dotnet_service/Database/Seeds/UserSeeder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using DotnetService.Database.Models;
using Microsoft.EntityFrameworkCore;
using Newtonsoft.Json;
using BCrypt.Net;

namespace DotnetService.Database.Seeds
{
    /// <summary>
    /// User seeder that creates users, optionally loading from system.admin.seed.json.
    /// Similar to the Python user_seeder.py implementation.
    /// </summary>
    public class UserSeeder : BaseSeeder
    {
        public UserSeeder(DatabaseContext context, int batchSize = 1000, bool verbose = true)
            : base(context, batchSize, verbose)
        {
        }
        
        private string HashPassword(string password)
        {
            return BCrypt.Net.BCrypt.HashPassword(password);
        }
        
        private Dictionary<string, object>? LoadSeedData()
        {
            var seedPath = Path.Combine(Directory.GetCurrentDirectory(), "seed.data", "system.admin.seed.json");
            
            try
            {
                if (File.Exists(seedPath))
                {
                    var json = File.ReadAllText(seedPath);
                    var data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
                    
                    if (data != null)
                    {
                        var result = new Dictionary<string, object>
                        {
                            { "username", data.GetValueOrDefault("UserName", "admin") ?? "admin" },
                            { "firstName", data.GetValueOrDefault("FirstName", "System") ?? "System" },
                            { "lastName", "Admin" },
                            { "email", data.GetValueOrDefault("Email", "sys.admin@example.com") ?? "sys.admin@example.com" },
                            { "password", data.GetValueOrDefault("Password", "ChangeMe123!") ?? "ChangeMe123!" },
                            { "phoneCode", data.GetValueOrDefault("PhoneCode", "+91") ?? "+91" }
                        };
                        
                        if (data.TryGetValue("PhoneNumber", out var phoneNumber) && phoneNumber != null)
                        {
                            result["phoneNumber"] = phoneNumber;
                        }
                        
                        return result;
                    }
                }
            }
            catch (Exception e)
            {
                if (Verbose)
                {
                    Console.WriteLine($"[WARN] Could not load seed data: {e.Message}");
                }
            }
            
            return null;
        }
        
        public override async Task<int> SeedAsync(int count = 2)
        {
            Console.WriteLine($"Seeding {count} users...");
            
            // Get all tenants
            var tenants = await Context.Tenants.ToListAsync();
            
            if (tenants.Count > 0)
            {
                Console.WriteLine($"[INFO] Assigning users to {tenants.Count} tenant(s)");
            }
            else
            {
                Console.WriteLine("[WARN] No tenants found. Users will be created without tenantId.");
            }
            
            // Load seed data
            var seedData = LoadSeedData();
            
            var defaultPassword = HashPassword("password123");
            
            // Default values
            var usernames = new List<string> { "john.smith", "sarah.brown", "mike.johnson", "jane.williams", "david.davis" };
            var firstNames = new List<string> { "John", "Jane", "Mike", "Sarah", "David" };
            var lastNames = new List<string> { "Smith", "Johnson", "Williams", "Brown", "Davis" };
            var emailDomain = "example.com";
            
            // Override with seed data if available
            if (seedData != null)
            {
                if (seedData.ContainsKey("username"))
                {
                    usernames = new List<string> { seedData["username"].ToString()! };
                }
                if (seedData.ContainsKey("firstName"))
                {
                    firstNames = new List<string> { seedData["firstName"].ToString()! };
                }
                if (seedData.ContainsKey("email") && seedData["email"].ToString()!.Contains("@"))
                {
                    emailDomain = seedData["email"].ToString()!.Split('@')[1];
                }
            }
            
            var users = new List<User>();
            var random = new Random();
            
            for (int i = 0; i < count; i++)
            {
                User user;
                
                // Use seed data for first user if available
                if (i == 0 && seedData != null)
                {
                    user = new User
                    {
                        Id = Guid.NewGuid(),
                        Username = seedData.GetValueOrDefault("username", $"user{i}").ToString()!,
                        Email = seedData.GetValueOrDefault("email", $"user{i}@{emailDomain}").ToString()!,
                        Password = HashPassword(seedData.GetValueOrDefault("password", "ChangeMe123!").ToString()!),
                        FirstName = (seedData.GetValueOrDefault("firstName", GetRandomChoice(firstNames) ?? "System") ?? "System").ToString(),
                        LastName = seedData.GetValueOrDefault("lastName", "Admin").ToString(),
                        IsActive = true,
                        IsEmailVerified = true,
                        IsPhoneVerified = false,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow
                    };
                    
                    if (seedData.ContainsKey("phoneNumber") && seedData["phoneNumber"] != null)
                    {
                        user.PhoneNumber = seedData["phoneNumber"].ToString();
                    }
                    if (seedData.ContainsKey("phoneCode") && seedData["phoneCode"] != null)
                    {
                        user.CountryCode = seedData["phoneCode"].ToString();
                    }
                }
                else
                {
                    user = new User
                    {
                        Id = Guid.NewGuid(),
                        Username = i < usernames.Count ? GetRandomChoice(usernames)! : $"user{i}",
                        Email = $"user{i}@{emailDomain}",
                        Password = defaultPassword,
                        FirstName = GetRandomChoice(firstNames),
                        LastName = GetRandomChoice(lastNames),
                        IsActive = GetRandomBoolean(0.9),
                        IsEmailVerified = GetRandomBoolean(0.8),
                        IsPhoneVerified = GetRandomBoolean(0.5),
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow
                    };
                }
                
                // Assign tenant if available
                if (tenants.Count > 0)
                {
                    var tenant = GetRandomChoice(tenants);
                    if (tenant != null)
                    {
                        user.TenantId = tenant.Id;
                    }
                }
                
                users.Add(user);
            }
            
            await BatchInsertAsync(users);
            Console.WriteLine($"Successfully seeded {CreatedCount} users");
            return CreatedCount;
        }
    }
}

